rosbag2_transport/include/rosbag2_transport/player_action_client.hpp

```diff
@@ -0,0 +1,179 @@
+  };
+
+  explicit
+  PlayerActionClient(
+    rclcpp_action::GenericClient::SharedPtr generic_client,
+    std::string & action_name,
+    const std::string & action_type,
+    rclcpp::Logger logger);
+
+  ~PlayerActionClient();
+
+  const std::string & get_action_name();
+
+  /// \brief Deserialize message to the type erased send goal service event
+  /// \param message - Serialized message
+  /// \return Shared pointer to the byte array with deserialized service event if success,
+  /// otherwise nullptr
+  std::shared_ptr<uint8_t[]> deserialize_send_goal_service_event(
+    const rcl_serialized_message_t & message);
+
+  /// \brief Deserialize message to the type erased cancel goal service event
+  /// \param message - Serialized message
+  /// \return Shared pointer to the byte array with deserialized service event if success,
+  /// otherwise nullptr
+  std::shared_ptr<uint8_t[]> deserialize_cancel_goal_service_event(
+    const rcl_serialized_message_t & message);
+
+  /// \brief Deserialize message to the type erased get result service event
+  /// \param message - Serialized message
+  /// \return Shared pointer to the byte array with deserialized service event if success,
+  /// otherwise nullptr
+  std::shared_ptr<uint8_t[]> deserialize_get_result_service_event(
+    const rcl_serialized_message_t & message);
+
+  /// \brief Extract the request data from the send_goal event message and asynchronously
+  // send the extracted request data.
+  void async_send_goal_request(
+    const std::shared_ptr<uint8_t[]> & type_erased_send_goal_service_event);
+
+  /// \brief Extract the request data from the cancel_goal event message and asynchronously
+  // send the extracted request data.
+  void async_send_cancel_request(
+    const std::shared_ptr<uint8_t[]> & type_erased_cancel_goal_service_event);
+
+  /// \brief Extract the request data from the get_result event message and asynchronously
+  // send the extracted request data.
+  void async_send_result_request(
+    const std::shared_ptr<uint8_t[]> & type_erased_get_result_service_event);
+
+  /// Wait until sent service requests will receive responses from service servers.
+  /// \param timeout - Timeout in fraction of seconds to wait for.
+  /// \return true if service requests successfully finished, otherwise false.
+  bool wait_for_sent_requests_to_finish(
+    std::chrono::duration<double> timeout = std::chrono::seconds(5));
+
+  ServiceEventType get_service_event_type(
+    const std::shared_ptr<uint8_t[]> & type_erased_service_event,
+    ServiceInterfaceInAction service_type_in_action);
+
+  rclcpp_action::GenericClient::SharedPtr generic_client()
+  {
+    return client_;
+  }
+
+  bool gold_handle_in_processing(const GoalID & goal_id);
+
+private:
+  rclcpp_action::GenericClient::SharedPtr client_;
+  std::string action_name_;
+  std::string action_type_;
+  const rclcpp::Logger logger_;
+
+  // Warning output once exceeding this value
```

Comment:
actually this prints every time it hits the limit.

```suggestion
  // Warning output when exceeding this value
```